core/src/main/java/org/apache/iceberg/rest/HttpRESTClient.java

```diff
@@ -0,0 +1,288 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.rest;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+import org.apache.hc.client5.http.classic.methods.HttpUriRequest;
+import org.apache.hc.client5.http.classic.methods.HttpUriRequestBase;
+import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
+import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
+import org.apache.hc.client5.http.impl.classic.HttpClients;
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.HttpHeaders;
+import org.apache.hc.core5.http.HttpStatus;
+import org.apache.hc.core5.http.Method;
+import org.apache.hc.core5.http.ParseException;
+import org.apache.hc.core5.http.impl.EnglishReasonPhraseCatalog;
+import org.apache.hc.core5.http.io.entity.EntityUtils;
+import org.apache.hc.core5.http.io.entity.StringEntity;
+import org.apache.hc.core5.io.CloseMode;
+import org.apache.iceberg.exceptions.RESTException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.rest.responses.ErrorResponse;
+import org.apache.iceberg.rest.responses.ErrorResponseParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An HttpClient for usage with the REST catalog.
+ */
+public class HttpRESTClient implements RESTClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HttpRESTClient.class);
+
+  private final String uri;
+  private final CloseableHttpClient httpClient;
+  private final ObjectMapper mapper;
+  private final Map<String, String> additionalHeaders;
+  private final Consumer<HttpUriRequest> requestInterceptor;
+
+  private HttpRESTClient(
+      String uri,
+      CloseableHttpClient httpClient,
+      ObjectMapper mapper,
+      Map<String, String> additionalHeaders,
+      Consumer<HttpUriRequest> requestInterceptor) {
+    this.uri = uri;
+    this.httpClient = httpClient != null ? httpClient : HttpClients.createDefault();
+    this.mapper = mapper != null ? mapper : new ObjectMapper();
+    this.additionalHeaders = additionalHeaders != null ? additionalHeaders : ImmutableMap.of();
+    this.requestInterceptor = requestInterceptor;
+  }
+
+  private static Optional<String> extractResponseBodyAsString(CloseableHttpResponse response, Class<?> responseType) {
+    try {
+      if (response.getEntity() == null) {
+        return Optional.empty();
+      }
+      return Optional.ofNullable(EntityUtils.toString(response.getEntity(), "UTF-8"));
```

Comment:
When would `EntityUtils.toString` return null?

Reply:
Okay, if `HttpEntity.getContent` returns null. This is probably worth a quick comment.

Also, newline between control flow and the following return statement.

Reply:
It seems like it can only return `null` in the case that `response.getEntity` is null.

Since we've already checked for that, I don't think we need to use `Optional.ofNullable` actually. Either one or the other.